docs/man/man3/Mail.3
.TH "Mail" 3 "22 Mar 2009" "Version 0.1" "m4php5" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Mail \- 
.SH SYNOPSIS
.br
.PP
Inherits \fBMaman\fP.
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fB__construct\fP ($config)"
.br
.ti -1c
.RI "\fBaddFile\fP ($filepath)"
.br
.ti -1c
.RI "\fBsetVars\fP ($vars)"
.br
.ti -1c
.RI "\fBassign\fP ($var, $val)"
.br
.ti -1c
.RI "\fBappend\fP ($var, $val)"
.br
.ti -1c
.RI "\fBsetTemplate\fP ($tpl)"
.br
.ti -1c
.RI "\fBsendToAdmin\fP ()"
.br
.ti -1c
.RI "\fBgetSubject\fP ()"
.br
.ti -1c
.RI "\fBgetAltbody\fP ()"
.br
.ti -1c
.RI "\fBgetBody\fP ()"
.br
.ti -1c
.RI "\fBfetch\fP ()"
.br
.ti -1c
.RI "\fBsendTo\fP ($mail)"
.br
.ti -1c
.RI "\fBrawsend\fP ($from, $to, $subject, $htmlbody, $altbody=null, $options=null, $attachments=null, $html=false)"
.br
.ti -1c
.RI "\fBgetContents\fP ()"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static & \fBfactory\fP ($template)"
.br
.in -1c
.SH "Detailed Description"
.PP 
\fBMail\fP class. Holds an \fBMtpl\fP object so mails can be handled just like \fBModule\fP templates Sent mails can be stored as log files (or not), mailed (or not), and also mailed to only one static recipient (for testing purpose) depending on the configuration options setup in the PEAR::getStaticProperty('Mail','global') 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Mail::__construct ($ config)"
.PP
Constructor
.PP
\fBParameters:\fP
.RS 4
\fI$config\fP string Template config 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "Mail::addFile ($ filepath)"
.PP
Attach files
.PP
public 
.PP
\fBParameters:\fP
.RS 4
\fI$filepath\fP string Path to file 
.RE
.PP

.SS "Mail::setVars ($ vars)"
.PP
Set mail vars
.PP
\fBParameters:\fP
.RS 4
\fI$vars\fP 
.RE
.PP

.SS "Mail::assign ($ var, $ val)"
.PP
Assign mail vars
.PP
public 
.PP
\fBParameters:\fP
.RS 4
\fI$var\fP string Var 
.br
\fI$val\fP string Value 
.RE
.PP

.SS "Mail::append ($ var, $ val)"
.PP
Append mail vars
.PP
public 
.PP
\fBParameters:\fP
.RS 4
\fI$var\fP string Var 
.br
\fI$val\fP string Value 
.RE
.PP

.SS "Mail::setTemplate ($ tpl)"
.PP
Set mail template
.PP
\fBParameters:\fP
.RS 4
\fI$tpl\fP string Template 
.RE
.PP

.SS "static& Mail::factory ($ template)\fC [static]\fP"
.PP
Email creation
.PP
public
.PP
\fBParameters:\fP
.RS 4
\fI$template\fP string Template 
.RE
.PP
\fBReturns:\fP
.RS 4
Email class 
.RE
.PP

.SS "Mail::sendToAdmin ()"
.PP
\fBMail\fP send to admin
.PP
public 
.SS "Mail::getSubject ()"
.PP
Get mail subject
.PP
public 
.PP
\fBReturns:\fP
.RS 4
Subject string 
.RE
.PP

.SS "Mail::getAltbody ()"
.PP
Get Alternate body
.PP
public 
.PP
\fBReturns:\fP
.RS 4
Body string 
.RE
.PP

.SS "Mail::getBody ()"
.PP
Get body
.PP
public 
.PP
\fBReturns:\fP
.RS 4
Body string 
.RE
.PP

.SS "Mail::fetch ()"
.PP
Fetch template
.PP
public 
.SS "Mail::sendTo ($ mail)"
.PP
Send to vars
.PP
public 
.PP
\fBParameters:\fP
.RS 4
\fI$mail\fP 
.RE
.PP

.SS "Mail::rawsend ($ from, $ to, $ subject, $ htmlbody, $ altbody = \fCnull\fP, $ options = \fCnull\fP, $ attachments = \fCnull\fP, $ html = \fCfalse\fP)"
.PP
Raw Send \fBMail\fP
.PP
public 
.PP
\fBParameters:\fP
.RS 4
\fI$from\fP string From 
.br
\fI$to\fP string To 
.br
\fI$subject\fP string Subject 
.br
\fI$htmlbody\fP string HTML Body 
.br
\fI$altbody\fP string Alternate body 
.br
\fI$options\fP array Options 
.br
\fI$attachments\fP array Attachments 
.br
\fI$html\fP boolean Is HTML 
.RE
.PP

.SS "Mail::getContents ()"
.PP
Get mail contents
.PP
public 
.PP
\fBReturns:\fP
.RS 4
Content 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for m4php5 from the source code.


Mail.php
<?php
/**
 * M PHP Framework
 *
 * @package      M
 * @subpackage   Mail
 * @author       Arnaud Sellenet <demental@sat2way.com>
 * @copyright    Copyright (c) 2003-2009 Arnaud Sellenet
 * @license      http://opensource.org/licenses/lgpl-license.php GNU Lesser General Public License
 * @version      0.1
 */

/**
 * 
 * Mail class. Holds an Mtpl object so mails can be handled just like Module templates
 * Sent mails can be stored as log files (or not), mailed (or not), and also mailed to only one static recipient (for testing purpose)
 * depending on the configuration options setup in the PEAR::getStaticProperty('Mail','global')
 *
 *
 */
class Mail extends Maman {

	/**
	 *
	 * Mail Template
	 *
	 * @var		string
	 * @access	public
	 */
	public $template;

	/**
	 *
	 * Attachments
	 *
	 * @var		array
	 * @access	public
	 */
	public $_attachments;

	/**
	 *
	 * View
	 *
	 * @var		class
	 * @access	public
	 */
	public $view;

	/**
	 *
	 * From
	 *
	 * @var		array
	 * @access	public
	 */
	public $from;

	/**
	 *
	 * To
	 *
	 * @var		array
	 * @access	public
	 */
	public $to;

	/**
	 *
	 * Constructor
	 *
	 * @param $config	string	Template config
	 */
	function __construct($config) {
		$this->setConfig($config);
		$this->view = new Mtpl($this->getConfig('template_dir'));

	}

	/**
	 *
	 * Attach files
	 *
	 * @access	public
	 * @param	$filepath	string	Path to file
	 */
	public function addFile($filepath)
	{
		$this->_attachments[]=$filepath;
	}

	/**
	 *
	 * Set mail vars
	 *
	 * @param $vars
	 */
	function setVars($vars) {
		$this->fetched = false;
		$this->view->assignArray($vars);
		if(key_exists('reply-to',$vars)) {
			$this->replyTo = $vars['reply-to'];
		}
	}

	/**
	 *
	 * Assign mail vars
	 *
	 * @access	public
	 * @param	$var	string	Var
	 * @param	$val	string	Value
	 */
	public function assign($var,$val)
	{
		$this->fetched = false;
		$this->view->assign($var,$val);
	}

	/**
	 *
	 * Append mail vars
	 *
	 * @access	public
	 * @param	$var	string	Var
	 * @param	$val	string	Value
	 */
	public function append($var,$val)
	{
		$this->fetched = false;
		$this->view->append($var,$val);
	}

	/**
	 *
	 * Set mail template
	 *
	 * @param $tpl		string	Template
	 */
	function setTemplate($tpl) {
		$this->template = $tpl;
	}

	/**
	 *
	 * Email creation
	 *
	 * @access	public
	 * @static
	 * @param	$template	string	Template
	 * @return	Email		class
	 */
	public static function &factory ( $template )
	{
		$opt = array('all'=>PEAR::getStaticProperty('Mail','global'));

		$mail = new Mail($opt);
		$mail->setConfigValue('template',$template);
		return $mail;
	}

	/**
	 *
	 * Mail send to admin
	 *
	 * @access public
	 */
	public function sendToAdmin ()
	{
		return $this->sendTo($this->getConfig('from'),$this->getConfig('fromname'));
	}

	/**
	 *
	 * Get mail subject
	 *
	 * @access	public
	 * @return	Subject		string
	 */
	public function getSubject()
	{
		if($this->fetched) {
			return $this->subject;
		}
		$this->fetch();
		return $this->subject;
	}

	/**
	 *
	 * Get Alternate body
	 *
	 * @access	public
	 * @return	Body	string
	 */
	public function getAltbody()
	{
		if($this->fetched) {
			return $this->altbody;
		}
		$this->fetch();
		return $this->altbody;
	}

	/**
	 *
	 * Get body
	 *
	 * @access	public
	 * @return	Body	string
	 */
	public function getBody()
	{
		if($this->fetched) {
			return $this->body;
		}
		$this->fetch();
		return $this->body;
	}

	/**
	 *
	 * Fetch template
	 *
	 * @access public
	 */
	public function fetch()
	{
		if($this->fetched) {return;}
		$this->body = trim($this->view->fetch($this->getConfig('template')));
		$this->altbody = $this->view->getCapture('altbody');
		if(empty($this->altbody)) {
			$this->altbody = $this->view->getCapture('alternate');
		}
		$this->subject = trim($this->view->getCapture('subject'));
		$this->fetched = 1;
	}

	/**
	 *
	 * Send to vars
	 *
	 * @access	public
	 * @param	$mail
	 */
	public function sendTo ($mail)
	{
		if(empty($this->from)) {
			$from = $this->from = array($this->getConfig('from'),$this->getConfig('fromname'));
		}
		$from = $this->from;
		$to=$mail;
		$this->fetch();
		if($this->altbody) {
			$html = true;
		} else {
			$html = false;
		}
		// TODO options, attachments
		$options = array();
		if($this->replyTo) {
			$options['reply-to'] = $this->replyTo;
		}
		$attachments = null;
		if(is_array($to)){
			foreach ($to as $e){
				$this->_sendmail($from,$e,$this->subject,$this->body,$this->altbody,$options,$this->_attachments,$html);
			}
		} else {
			$this->_sendmail($from,$to,$this->subject,$this->body,$this->altbody,$options,$this->_attachments,$html);
		}
	}

	/**
	 *
	 * Raw Send Mail
	 *
	 * @access public
	 * @param $from			string	From
	 * @param $to			string	To
	 * @param $subject		string	Subject
	 * @param $htmlbody		string	HTML Body
	 * @param $altbody		string	Alternate body
	 * @param $options		array	Options
	 * @param $attachments	array	Attachments
	 * @param $html			boolean	Is HTML
	 */
	public function rawsend($from,$to,$subject,$htmlbody,$altbody=null,$options = null,$attachments=null,$html=false)
	{
		$this->_sendmail($from,$to,$subject,$htmlbody,$altbody,$options,$attachments,$html);
	}

	/**
	 *
	 * Get mail contents
	 *
	 * @access 	public
	 * @return 	Content
	 */
	public function getContents()
	{
		$out = new StdClass();
		$out->body=$this->view->fetch($this->getConfig('template'));
		$out->subject = $this->view->getCapture('subject');
		$out->altbody = $this->view->getCapture('altbody');


		return $out;
	}

	/**
	 *
	 * Send Mail
	 *
	 * @access private
	 * @param $from			string	From
	 * @param $to			string	To
	 * @param $subject		string	Subject
	 * @param $body			string	HTML Body
	 * @param $altbody		string	Alternate body
	 * @param $options		array	Options
	 * @param $attachments	array	Attachments
	 * @param $html			boolean	Is HTML
	 */
	private function _sendmail($from,$to,$subject,$body,$altbody = null, $options = null, $attachments = null, $html = false) {

		if(!is_array($from)){
			$from=array($from,$from);
		}
		if($only = $this->getConfig('sendonlyto')){
			$note="-------Note : Mode test activé, ce message était normalement destiné à ".$to."---------\r\n";
			$to=$only;
		}
		if($this->getConfig('sendmail')){
			$mail = new phpmailer();
			$mail->PluginDir='M/lib/phpmailer/';
			if($this->getConfig('smtp')) {
				$mail->isSMTP();
				$mail->Host=$this->getConfig('smtphost');
			}
			$mail->CharSet=$this->getConfig('encoding');
			$mail->AddAddress($to);                // Adresse email de reception
			$mail->Subject = $subject;                    // Sujet
			$mail->Body = $note.$body;          // Corps du message
			$mail->AltBody=$altbody;
			if(!is_array($from)){
				$from=array($from,$from);
			}
			$mail->From = $from[0];              // Adresse email de l'expediteur (optionnel)
			$mail->FromName = $from[1];

			if(key_exists('reply-to',$options)){
				$mail->AddReplyTo($options['reply-to']);
				unset($options['reply-to']);
			}
			if(key_exists('Sender',$options)) {
				$mail->Sender = $options['Sender'];
			}

			if(null!=$attachments){
				if(!is_array($attachments)){
					$attachments=array($attachments);
				}
				foreach($attachments as $k=>$v){
					if(!$mail->AddAttachment($v, basename($v))){
						echo "Erreur impossible d'ajouter pièce jointe $v";
					}
				}
			}
			$mail->IsHTML($html);
			$result = 	$mail->send();
		}
		/**
		 * Enregistrement en logfile
		 */
		if($this->getConfig('logmail')){
			$newmailFile=date("Y-m-d-H-i-s").".mlog";
			$fp=fopen($this->getConfig('log_folder').$newmailFile,"a+");
			fwrite($fp,"Message de : ".$from[1]."<".$from[0].">\r\n");
			fwrite($fp,"à : ".$to."\r\n");
			fwrite($fp,"Sujet : ".stripslashes($subject)."\r\n");
			fwrite($fp,"Message : ".stripslashes($note.$body)."\r\n");
			fwrite($fp,"\r\n");
			if(!is_array($attachments)){
				$attachments=array($attachments);
			}
			foreach($attachments as $k=>$v){
				fwrite($fp,"Pièce jointe : ".basename($v)."\r\n");
			}
			if($this->getConfig('sendmail')){
				fwrite($fp,"Ce message a été envoyé par mail.\r\n----------------------------------------\r\n\r\n");
			}
			fclose($fp);
		}
	}
}